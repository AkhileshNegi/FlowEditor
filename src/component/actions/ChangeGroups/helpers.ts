import { Group, ChangeGroups } from '../../../flowTypes';
import { SearchResult } from '../../../store';

export const mapGroupsToSearchResults = (groups: Group[]) =>
    groups.map(({ name, uuid }) => ({ name, id: uuid }));

export const mapSearchResultsToGroups = (searchResults: SearchResult[]) =>
    searchResults.map(result => ({
        uuid: result.id,
        name: result.name
    }));

export const getGroups = (action: ChangeGroups): SearchResult[] => {
    if (action.groups == null) {
        return [];
    }

    if (action.groups.length && action.type !== 'add_contact_groups') {
        return mapGroupsToSearchResults(action.groups);
    }

    return [];
};
