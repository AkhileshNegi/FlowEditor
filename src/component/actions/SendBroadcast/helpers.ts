import { BroadcastMsg, Contact, Group } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';

export const getRecipients = (action: BroadcastMsg): Asset[] => {
    const selected = action.groups.map((group: Group) => {
        return { id: group.uuid, name: group.name, type: AssetType.Group };
    });

    return selected.concat(
        action.contacts.map((contact: Contact) => {
            return { id: contact.uuid, name: contact.name, type: AssetType.Contact };
        })
    );
};
