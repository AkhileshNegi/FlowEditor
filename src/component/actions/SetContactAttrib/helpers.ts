import { Types } from '../../../config/typeConfigs';
import { Action, SetContactField, SetContactName, SetContactProperty } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';
import { snakify, titleCase } from '../../../utils';

export type ContactProperty = Types.set_contact_name;

export const newFieldAction = ({
    uuid,
    value,
    name
}: {
    uuid: string;
    value: string;
    name: string;
}): SetContactField => ({
    type: Types.set_contact_field,
    field: {
        key: snakify(name),
        name: titleCase(name)
    },
    uuid,
    value
});

export const newPropertyAction = ({
    uuid,
    value,
    type
}: {
    uuid: string;
    value: string;
    type: AssetType;
}): SetContactProperty => {
    const action: Partial<Action> = {
        uuid
    };
    switch (type) {
        case AssetType.Name:
            return {
                ...action,
                type: Types.set_contact_name,
                name: value
            } as SetContactName;
    }
};

export const fieldToAsset = ({ field: { key, name } }: SetContactField): Asset => ({
    id: key,
    name,
    type: AssetType.Field
});

export const propertyToAsset = (property: SetContactProperty): Asset => {
    switch (property.type) {
        case Types.set_contact_name:
            return {
                type: AssetType.Name,
                name: titleCase(AssetType.Name),
                id: AssetType.Name
            };
    }
};
