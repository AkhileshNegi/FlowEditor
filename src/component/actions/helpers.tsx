import * as React from 'react';

import { Asset, AssetType } from '../../services/AssetService';

const styles = require('../shared.scss');

export const renderAssetList = (assets: Asset[], max: number): JSX.Element[] => {
    return assets.reduce((elements, asset, idx) => {
        if (idx <= max - 1) {
            elements.push(renderAsset(asset));
        } else if (idx > max - 1 && elements.length === max) {
            elements.push(<div key="ellipses">...</div>);
        }
        return elements;
    }, []);
};

export const renderAsset = (asset: Asset) => {
    switch (asset.type) {
        case AssetType.Group:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeGroup} icn-group`} />
                    {asset.name}
                </div>
            );
        case AssetType.Label:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeLabel} icn-label`} />
                    {asset.name}
                </div>
            );
    }

    return (
        <div className={styles.nodeAsset} key={asset.id}>
            {asset.name}
        </div>
    );
};
