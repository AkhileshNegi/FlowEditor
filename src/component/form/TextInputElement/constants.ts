import { CompletionOption } from '../../../store';

// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
export enum KeyValues {
    KEY_AT = '@',
    KEY_SPACE = ' ',
    KEY_ENTER = 'Enter',
    KEY_UP = 'ArrowUp',
    KEY_DOWN = 'ArrowDown',
    KEY_TAB = 'Tab',
    KEY_ESC = 'Escape',
    KEY_BACKSPACE = 'Backspace',
    KEY_N = 'n',
    KEY_P = 'p'
}

export enum TopLevelVariables {
    contact = 'contact',
    input = 'input',
    run = 'run',
    parent = 'parent',
    child = 'child',
    trigger = 'trigger'
}

export const MAX_GSM_SINGLE = 160;
export const MAX_GSM_MULTI = 153;
export const MAX_UNICODE_SINGLE = 70;
export const MAX_UNICODE_MULTI = 67;

export const COMPLETION_HELP = 'Tab to complete, enter to select';

export const getFlowOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.flow`, description: `The flow in which a ${accessor} run takes place` },
        {
            name: `${prefix}.flow.uuid`,
            description: `The UUID of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.name`,
            description: `The name of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.revision`,
            description: `The revision number of the flow in which a ${accessor} run takes place`
        }
    ];
};

export const getWebhookOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        {
            name: `${prefix}.webhook`,
            description: `The body of the response to the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status`,
            description: `The status of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status_code`,
            description: `The status code returned from the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.url`,
            description: `The URL that was called by the last webhook request in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.body`,
            description: `The body of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.json`,
            description: `The JSON parsed body of the response to the last webhook request in a ${accessor} run, can access subelements`
        },
        {
            name: `${prefix}.webhook.request`,
            description: `The raw request last made in a ${accessor} run, including headers`
        },
        {
            name: `${prefix}.webhook.response`,
            description: `The raw response last received in a ${accessor} run, including headers`
        },
        { name: `${prefix}.results`, description: `Results collected in a ${accessor} run` }
    ];
};

export const getInputOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.input`, description: `A ${accessor} run's last input` },
        { name: `${prefix}.input.uuid`, description: `The UUID of a ${accessor} run's last input` },
        { name: `${prefix}.input.type`, description: `The type of a ${accessor} run's last input` },
        {
            name: `${prefix}.input.channel`,
            description: `The channel a ${accessor} run's last input was received on`
        },
        {
            name: `${prefix}.input.created_on`,
            description: `The time a ${accessor} run's last input was created`
        },
        {
            name: `${prefix}.input.text`,
            description: `The text of a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.attachments`,
            description: `The attachments on a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.urn`,
            description: `The URN a ${accessor} run's last input was received on`
        }
    ];
};

export const getContactOptions = (accessor?: string) => {
    const prefix = accessor ? `${accessor}.` : '';
    const descriptor = accessor
        ? accessor === TopLevelVariables.run ? `${accessor}'s` : `${accessor} run's`
        : '';
    return [
        { name: `${prefix}contact`, description: `The name of the ${descriptor} contact` },
        { name: `${prefix}contact.name`, description: `The name of the ${descriptor} contact` },
        {
            name: `${prefix}contact.first_name`,
            description: `The first name of the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.language`,
            description: `The language code for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.fields`,
            description: `Custom fields on the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.groups`,
            description: `The groups the ${descriptor} contact is a member of`
        },
        { name: `${prefix}contact.urns`, description: `URNs on the ${descriptor} contact` },
        {
            name: `${prefix}contact.urns.tel`,
            description: `The preferred telephone number for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.urns.telegram`,
            description: `The preferred telegram id for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.channel`,
            description: `The ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.uuid`,
            description: `The UUID of the ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.name`,
            description: `The name of the ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.address`,
            description: `The address of the ${descriptor} contact's preferred channel`
        }
    ];
};

export const RUN_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.run, description: 'A run in this flow' },
    ...getFlowOptions(),
    ...getContactOptions(TopLevelVariables.run),
    ...getInputOptions(),
    ...getWebhookOptions()
];

export const CHILD_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.child, description: 'Run details collected in a child flow, if any' },
    ...getFlowOptions(TopLevelVariables.child),
    ...getContactOptions(TopLevelVariables.child),
    ...getInputOptions(TopLevelVariables.child),
    ...getWebhookOptions(TopLevelVariables.child)
];

export const PARENT_OPTIONS: CompletionOption[] = [
    {
        name: TopLevelVariables.parent,
        description: 'Run details collected by a parent flow, if any'
    },
    ...getFlowOptions(TopLevelVariables.parent),
    ...getContactOptions(TopLevelVariables.parent),
    ...getInputOptions(TopLevelVariables.parent),
    ...getWebhookOptions(TopLevelVariables.parent)
];

export const TRIGGER_OPTIONS: CompletionOption[] = [
    { name: 'trigger', description: 'The trigger that initiated the session' },
    { name: 'trigger.type', description: 'The type of the trigger, one of “manual” or “flow”' },
    { name: 'trigger.params', description: 'The parameters passed to the trigger' }
];

export const OPTIONS: CompletionOption[] = [
    ...getContactOptions(),
    ...RUN_OPTIONS,
    ...CHILD_OPTIONS,
    ...PARENT_OPTIONS,
    ...TRIGGER_OPTIONS
];

export const TOP_LEVEL_OPTIONS = OPTIONS.filter(
    ({ name }) =>
        name === TopLevelVariables.contact ||
        name === TopLevelVariables.input ||
        name === TopLevelVariables.run ||
        name === TopLevelVariables.parent ||
        name === TopLevelVariables.child ||
        name === TopLevelVariables.trigger
);

export const GSM: { [key: string]: number } = {
    // char: charCode
    0: 48,
    1: 49,
    2: 50,
    3: 51,
    4: 52,
    5: 53,
    6: 54,
    7: 55,
    8: 56,
    9: 57,
    '\n': 10,
    '\f': 12,
    '\r': 13,
    ' ': 32,
    '!': 33,
    '"': 34,
    '#': 35,
    $: 36,
    '%': 37,
    '&': 38,
    "'": 39,
    '(': 40,
    ')': 41,
    '*': 42,
    '+': 43,
    ',': 44,
    '-': 45,
    '.': 46,
    '/': 47,
    ':': 58,
    ';': 59,
    '<': 60,
    '=': 61,
    '>': 62,
    '?': 63,
    '@': 64,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    '[': 91,
    '\\': 92,
    ']': 93,
    '^': 94,
    _: 95,
    a: 97,
    b: 98,
    c: 99,
    d: 100,
    e: 101,
    f: 102,
    g: 103,
    h: 104,
    i: 105,
    j: 106,
    k: 107,
    l: 108,
    m: 109,
    n: 110,
    o: 111,
    p: 112,
    q: 113,
    r: 114,
    s: 115,
    t: 116,
    u: 117,
    v: 118,
    w: 119,
    x: 120,
    y: 121,
    z: 122,
    '{': 123,
    '|': 124,
    '}': 125,
    '~': 126,
    '¡': 161,
    '£': 163,
    '¤': 164,
    '¥': 165,
    '§': 167,
    '¿': 191,
    Ä: 196,
    Å: 197,
    Æ: 198,
    Ç: 199,
    É: 201,
    Ñ: 209,
    Ö: 214,
    Ø: 216,
    Ü: 220,
    ß: 223,
    à: 224,
    ä: 228,
    å: 229,
    æ: 230,
    è: 232,
    é: 233,
    ì: 236,
    ñ: 241,
    ò: 242,
    ö: 246,
    ø: 248,
    ù: 249,
    ü: 252,
    Γ: 915,
    Δ: 916,
    Θ: 920,
    Λ: 923,
    Ξ: 926,
    Π: 928,
    Σ: 931,
    Φ: 934,
    Ψ: 936,
    Ω: 937,
    '€': 8364
};
