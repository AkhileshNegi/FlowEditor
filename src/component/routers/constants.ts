export const WAIT_LABEL = 'If the message response...';
export const EXPRESSION_LABEL = 'If the expression...';
export const GROUP_LABEL = "Select the group(s) you'd like to split by below";
export const OPERAND_LOCALIZATION_DESC =
    'Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.';
