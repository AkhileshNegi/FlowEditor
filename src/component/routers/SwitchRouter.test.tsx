import update from 'immutability-helper';
import * as React from 'react';
import { CaseElementProps } from '~/component/form/CaseElement';
import TextInputElement from '~/component/form/TextInputElement';
import { DEFAULT_OPERAND } from '~/component/NodeEditor';
import { getOperatorConfig, Operators } from '~/config/operatorConfigs';
import { getTypeConfig, Types } from '~/config/typeConfigs';
import { SwitchRouter } from '~/flowTypes';
import { AssetType } from '~/services/AssetService';
import { getLocalizations } from '~/store/helpers';
import { composeComponentTestUtils, getSpecWrapper, setMock } from '~/testUtils';
import { createCase, createExit, createWaitRouterNode, Spanish } from '~/testUtils/assetCreators';
import { reorderList } from '~/utils';

import { EXPRESSION_LABEL } from './constants';
import {
    addFocus,
    casePropsFromElement,
    DragCursor,
    getItemStyle,
    getListStyle,
    InputToFocus,
    leadInSpecId,
    SwitchRouterForm,
    SwitchRouterFormProps
} from './SwitchRouter';

let mockUuidCounts = 1;
jest.mock('uuid', () => {
    return {
        v4: jest.fn(() => `generated_uuid_${mockUuidCounts++}`)
    };
});

const rulesMeta = [
    { exitUUID: 'exit-1', exitName: 'Yes', type: Operators.has_any_word, args: ['y, yes'] },
    { exitUUID: 'exit-2', exitName: 'No', type: Operators.has_any_word, args: ['n, no'] },
    {
        exitUUID: 'exit-3',
        exitName: 'No Response',
        type: Operators.has_wait_timed_out,
        args: ['@run']
    }
];

const exits = rulesMeta.map(({ exitUUID, exitName }, idx) =>
    createExit({
        uuid: exitUUID,
        name: exitName,
        destination_node_uuid: `node-${idx}`
    })
);

const cases = rulesMeta.map(({ exitUUID, type, args }, idx) =>
    createCase({
        uuid: `case-${idx}`,
        type,
        exit_uuid: exitUUID,
        args
    })
);

const nodeToEdit = createWaitRouterNode({
    exits,
    cases,
    timeout: 300
});

const baseProps: SwitchRouterFormProps = {
    language: { id: 'eng', name: 'English', type: AssetType.Language },
    typeConfig: getTypeConfig(Types.wait_for_response),
    translating: false,
    settings: { originalNode: nodeToEdit },
    operand: DEFAULT_OPERAND,
    showAdvanced: false,
    saveLocalizations: jest.fn(),
    updateRouter: jest.fn(),
    onExpressionChanged: jest.fn(),
    getExitTranslations: jest.fn(),
    getResultNameField: jest.fn(),
    onBindWidget: jest.fn(),
    onBindAdvancedWidget: jest.fn(),
    removeWidget: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(SwitchRouterForm, baseProps);

describe(SwitchRouterForm.name, () => {
    beforeEach(() => {
        mockUuidCounts = 1;
    });

    describe('helpers', () => {
        describe('getListStyle', () => {
            it('should return null', () => {
                expect(getListStyle(true, true)).toBeNull();
            });

            it('should return dragging cursor', () => {
                expect(getListStyle(true, false)).toEqual({ cursor: DragCursor.move });
            });

            it('should return pointer cursor', () => {
                expect(getListStyle(false, false)).toEqual({ cursor: DragCursor.pointer });
            });
        });

        describe('getItemStyle', () => {
            const draggableStyle = {
                position: 'fixed',
                boxSizing: 'border-box',
                pointerEvents: 'none',
                zIndex: 4500,
                width: 605,
                height: 40,
                top: 262,
                left: 389,
                margin: 0,
                transform: 'translate(0px, 0.036933467580315696px)',
                WebkitTouchCallout: 'none',
                WebkitTapHighlightColor: 'rgba(0,0,0,0)',
                touchAction: 'manipulation'
            };

            it('should return draggable style', () => {
                expect(getItemStyle(draggableStyle, false)).toEqual({
                    ...draggableStyle,
                    userSelect: 'none',
                    outline: 'none'
                });
            });

            it('should return modified draggable style', () => {
                expect(getItemStyle(draggableStyle, true)).toMatchSnapshot();
            });
        });

        describe('addFocus', () => {
            it('should apply arg focus', () => {
                expect(addFocus({} as any, InputToFocus.args)).toEqual({ focusArgs: true });
            });

            it('should apply min focus', () => {
                expect(addFocus({} as any, InputToFocus.min)).toEqual({ focusMin: true });
            });

            it('should apply max focus', () => {
                expect(addFocus({} as any, InputToFocus.max)).toEqual({ focusMax: true });
            });

            it('should apply exit focus', () => {
                expect(addFocus({} as any, InputToFocus.exit)).toEqual({ focusExit: true });
            });
        });

        describe('casePropsFromElement', () => {
            it('should convert CaseElement properties into a CaseElementProps object', () => {
                expect(
                    casePropsFromElement({
                        caseElement: {
                            props: {
                                kase: {
                                    uuid: '75541df6-f5f7-4f7b-944b-9604cd6cf338',
                                    exit_uuid: null
                                }
                            },
                            state: {
                                operatorConfig: getOperatorConfig(Operators.has_any_word),
                                arguments: ['no'],
                                exitName: 'No'
                            }
                        } as any
                    })
                ).toMatchSnapshot();
            });
        });
    });

    describe('render', () => {
        it('should render base form', () => {
            const { wrapper, props } = setup(false, { getResultNameField: setMock() });

            expect(wrapper.find('CaseElement').length).toBe(3);
            expect(props.getResultNameField).toHaveBeenCalledTimes(1);
        });

        it('should render exit translation form', () => {
            const { wrapper, props } = setup(true, {
                translating: { $set: true },
                getExitTranslations: setMock(() => <div />)
            });

            expect(props.getExitTranslations).toHaveBeenCalledTimes(1);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render advanced form', () => {
            const renderAdvancedSpy = spyOn('renderAdvanced');
            const translations = {
                [(nodeToEdit.router as SwitchRouter).cases[0].uuid]: {
                    name: ['Si'],
                    arguments: ['si, simÃ³n']
                }
            };
            const localizations = getLocalizations(nodeToEdit, null, Spanish, translations);
            const { wrapper, props } = setup(true, {
                translating: { $set: true },
                showAdvanced: { $set: true },
                $merge: {
                    settings: {
                        originalNode: nodeToEdit,
                        localizations
                    }
                }
            });

            expect(renderAdvancedSpy).toHaveBeenCalledTimes(1);
            expect(wrapper).toMatchSnapshot();

            renderAdvancedSpy.mockRestore();
        });
    });

    describe('instance methods', () => {
        describe('getCaseContext', () => {
            it('should return unwrapped, empty case', () => {
                const { wrapper, props } = setup(true, {
                    settings: {
                        originalNode: { router: { $merge: { cases: [] } } }
                    }
                });

                expect(getSpecWrapper(wrapper, leadInSpecId).exists()).toBeTruthy();
                expect(wrapper.find('CaseElement').length).toBe(1);
                expect(wrapper).toMatchSnapshot();
            });
        });

        describe('getLeadIn', () => {
            it('should return expression input', () => {
                const { wrapper, props } = setup(true, {
                    typeConfig: { $set: getTypeConfig(Types.split_by_expression) }
                });

                expect(wrapper.find('p').text()).toBe(EXPRESSION_LABEL);
                expect(wrapper.find(TextInputElement).exists()).toBeTruthy();
                expect(getSpecWrapper(wrapper, leadInSpecId)).toMatchSnapshot();
            });
        });

        describe('getCasesToRender', () => {
            const partialCase: CaseElementProps = {
                exitName: '',
                kase: {
                    arguments: [],
                    exit_uuid: null,
                    type: Operators.has_all_words,
                    uuid: 'partial-case'
                }
            };

            it('should add an undraggable, displayable case', () => {
                const { wrapper } = setup(true, {
                    settings: {
                        originalNode: {
                            router: {
                                $merge: {
                                    cases: [cases[0]]
                                }
                            }
                        }
                    }
                });

                expect(wrapper).toMatchSnapshot();
            });

            it('should add partially-formed case', () => {
                const { wrapper, props } = setup(false);

                // create a partially-formed case
                wrapper.setState({
                    displayableCases: [...wrapper.state('displayableCases')[0], partialCase]
                });

                // register update
                wrapper.update();

                // find our cases
                const caseElements = wrapper.find('CaseElement');

                expect(caseElements.length).toBe(2);
                expect(caseElements.at(1).props()).toMatchSnapshot();
            });

            it("should hold off on adding an empty case if displayable case's operator expects 1 or more operand and its arguments and exitName are empty", () => {
                const { wrapper, props } = setup(false);

                // create a partially-formed case
                wrapper.setState({
                    displayableCases: [...wrapper.state('displayableCases'), partialCase]
                });

                // register update
                wrapper.update();

                // find our cases
                const caseElements = wrapper.find('CaseElement');

                expect(caseElements.length).toBe(3);
                expect(caseElements.at(2).props()).toMatchSnapshot();
            });
        });

        describe('onDragEnd', () => {
            it('should return early if no destination is passed', () => {
                const { wrapper, instance } = setup();

                expect(instance.onDragEnd({})).toBeUndefined();
            });

            it('should update state w/ reordered cases', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance } = setup();
                const mockDropResult = {
                    draggableId: '3ed694c3-9f51-4b2e-b459-92e3e15c8bef',
                    type: 'DEFAULT',
                    source: {
                        index: 0,
                        droppableId: 'droppable'
                    },
                    destination: {
                        droppableId: 'droppable',
                        index: 2
                    }
                };
                const reorderedCases = reorderList(
                    wrapper.state('displayableCases'),
                    mockDropResult.source.index,
                    mockDropResult.destination.index
                );

                instance.onDragEnd(mockDropResult);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({ displayableCases: reorderedCases });

                setStateSpy.mockRestore();
            });
        });

        describe('onValid', () => {
            it('should update router', () => {
                const { wrapper, instance, props } = setup(true, {
                    updateRouter: setMock()
                });

                instance.onValid();

                expect(props.updateRouter).toHaveBeenCalledTimes(1);
                expect(props.updateRouter).toHaveBeenCalledWith(wrapper.state('displayableCases'));
            });

            it('should save localizations', () => {
                const widgets = {
                    ['7eb09740-7692-401c-ad41-41545bf442bc']: true,
                    ['8adcc59f-3390-4f34-b02f-4b0fbe6eefe3']: true
                };
                const { wrapper, instance, props } = setup(true, {
                    translating: { $set: true },
                    saveLocalizations: setMock()
                });

                instance.onValid(widgets);

                expect(props.saveLocalizations).toHaveBeenCalledTimes(1);
                expect(props.saveLocalizations).toHaveBeenCalledWith(
                    widgets,
                    wrapper.state('displayableCases')
                );
            });
        });

        describe('handleCaseRemoved', () => {
            it('should update state, remove widget', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance, props } = setup(true, {
                    removeWidget: setMock()
                });
                const { uuid: caseUUID } = (props.settings.originalNode
                    .router as SwitchRouter).cases[0];
                const caseToRemove = {
                    props: {
                        name: `case_${caseUUID}`,
                        kase: {
                            uuid: caseUUID
                        }
                    }
                };

                const displayableCases = update(wrapper.state('displayableCases'), {
                    $splice: [[0, 1]]
                });

                instance.handleCaseRemoved(caseToRemove);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith(
                    { displayableCases },
                    expect.any(Function)
                );
                expect(props.removeWidget).toHaveBeenCalledTimes(1);
                expect(props.removeWidget).toHaveBeenCalledWith(caseToRemove.props.name);

                setStateSpy.mockRestore();
            });
        });

        describe('handleCaseChanged', () => {
            it('should update existing case', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance } = setup();
                const { uuid: existingCaseUUID, type: operatorType } = cases[0];
                const { uuid: exitsingExitUUID, name: exitName } = exits[0];
                // updating the first case's arguments
                const updatedCase = {
                    props: {
                        kase: {
                            uuid: existingCaseUUID,
                            exit_uuid: exitsingExitUUID
                        }
                    },
                    state: {
                        operatorConfig: getOperatorConfig(operatorType),
                        arguments: ['y', 'yes', 'sure'],
                        exitName
                    }
                };
                const updatedCaseProps = casePropsFromElement({
                    caseElement: updatedCase as any,
                    inputToFocus: InputToFocus.args
                });

                instance.handleCaseChanged(updatedCase, InputToFocus.args);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({
                    displayableCases: [
                        updatedCaseProps,
                        ...wrapper.state('displayableCases').slice(1)
                    ]
                });

                setStateSpy.mockRestore();
            });

            it('should add new case', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance } = setup();
                const newCase = {
                    props: {
                        kase: {
                            uuid: '75541df6-f5f7-4f7b-944b-9604cd6cf338',
                            exit_uuid: null as string
                        }
                    },
                    state: {
                        operatorConfig: getOperatorConfig(Operators.has_any_word),
                        arguments: ['maybe'],
                        exitName: 'Maybe'
                    }
                };
                const newCaseProps = casePropsFromElement({
                    caseElement: newCase as any,
                    inputToFocus: InputToFocus.args
                });
                const currentState = wrapper.state('displayableCases');

                instance.handleCaseChanged(newCase, InputToFocus.args);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({
                    displayableCases: [
                        ...currentState.slice(0, currentState.length - 1),
                        newCaseProps
                    ]
                });

                setStateSpy.mockRestore();
            });
        });
    });
});
