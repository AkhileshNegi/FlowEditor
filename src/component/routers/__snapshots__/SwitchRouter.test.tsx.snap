// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchRouterForm helpers casePropsFromElement should convert CaseElement properties into a CaseElementProps object 1`] = `
Object {
  "exitName": "No",
  "kase": Object {
    "arguments": Array [
      "no",
    ],
    "exit_uuid": null,
    "type": "has_any_word",
    "uuid": "75541df6-f5f7-4f7b-944b-9604cd6cf338",
  },
}
`;

exports[`SwitchRouterForm helpers getItemStyle should return modified draggable style 1`] = `
Object {
  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
  "WebkitTouchCallout": "none",
  "background": "#f2f9fc",
  "borderRadius": 4,
  "boxSizing": "border-box",
  "height": 55,
  "left": 20,
  "margin": 0,
  "opacity": 0.75,
  "outline": "none",
  "pointerEvents": "none",
  "position": "fixed",
  "top": 172,
  "touchAction": "manipulation",
  "transform": "translate(0px, 0.036933467580315696px)",
  "userSelect": "none",
  "width": 605,
  "zIndex": 4500,
}
`;

exports[`SwitchRouterForm instance methods getCaseContext should return unwrapped, empty case 1`] = `
<React.Fragment>
  <div
    className="instructions"
    data-spec="lead-in"
  >
    If the message response...
  </div>
  <React.Fragment>
    <CaseElement
      data-spec="case"
      empty={true}
      exitName=""
      kase={
        Object {
          "exit_uuid": null,
          "type": "has_any_word",
          "uuid": "generated_uuid_1",
        }
      }
      key="generated_uuid_1"
      name="case_generated_uuid_1"
      onChange={[Function]}
      onRemove={[Function]}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`SwitchRouterForm instance methods getCasesToRender should add an undraggable, displayable case 1`] = `
<React.Fragment>
  <div
    className="instructions"
    data-spec="lead-in"
  >
    If the message response...
  </div>
  <React.Fragment>
    <DragDropContext
      onDragEnd={[Function]}
    >
      <Connect(Droppable)
        droppableId="droppable"
      />
    </DragDropContext>
    <CaseElement
      data-spec="case"
      empty={true}
      exitName=""
      kase={
        Object {
          "exit_uuid": null,
          "type": "has_any_word",
          "uuid": "generated_uuid_1",
        }
      }
      key="generated_uuid_1"
      name="case_generated_uuid_1"
      onChange={[Function]}
      onRemove={[Function]}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`SwitchRouterForm instance methods getCasesToRender should add partially-formed case 1`] = `
Object {
  "data-spec": "case",
  "exitName": "",
  "focusArgs": undefined,
  "focusExit": undefined,
  "focusMax": undefined,
  "focusMin": undefined,
  "kase": Object {
    "arguments": Array [],
    "exit_uuid": null,
    "type": "has_all_words",
    "uuid": "partial-case",
  },
  "name": "case_partial-case",
  "onChange": [Function],
  "onRemove": [Function],
}
`;

exports[`SwitchRouterForm instance methods getCasesToRender should hold off on adding an empty case if displayable case's operator expects 1 or more operand and its arguments and exitName are empty 1`] = `
Object {
  "data-spec": "case",
  "exitName": "",
  "focusArgs": undefined,
  "focusExit": undefined,
  "focusMax": undefined,
  "focusMin": undefined,
  "kase": Object {
    "arguments": Array [],
    "exit_uuid": null,
    "type": "has_all_words",
    "uuid": "partial-case",
  },
  "name": "case_partial-case",
  "onChange": [Function],
  "onRemove": [Function],
}
`;

exports[`SwitchRouterForm instance methods getLeadIn should return expression input 1`] = `
<div
  className="instructions"
  data-spec="lead-in"
>
  <React.Fragment>
    <p>
      If the expression...
    </p>
    <Connect(TextInputElement)
      autocomplete={true}
      key="wait-router"
      name="Expression"
      onChange={[MockFunction]}
      required={true}
      showLabel={false}
      value="@run.input"
    />
  </React.Fragment>
</div>
`;

exports[`SwitchRouterForm render should render advanced form 1`] = `
<React.Fragment>
  <div
    className="translatingOperatorTitle"
    data-spec="advanced-title"
  >
    Rules
  </div>
  <div
    className="translatingOperatorInstructions"
    data-spec="advanced-instructions"
  >
    Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.
  </div>
  <div
    className="translating_operator_container"
    data-spec="operator-field"
    key="translate_case-0"
  >
    <div
      className="translating_operator"
      data-spec="verbose-name"
    >
      has any of the words
    </div>
    <div
      className="translating_from"
      data-spec="argument-to-translate"
    >
      y, yes
    </div>
    <div
      className="translating_to"
    >
      <Connect(TextInputElement)
        data-spec="translation-input"
        name="case-0"
        placeholder="English Translation"
        showLabel={false}
        value="si, simÃ³n"
      />
    </div>
  </div>
  <div
    className="translating_operator_container"
    data-spec="operator-field"
    key="translate_case-1"
  >
    <div
      className="translating_operator"
      data-spec="verbose-name"
    >
      has any of the words
    </div>
    <div
      className="translating_from"
      data-spec="argument-to-translate"
    >
      n, no
    </div>
    <div
      className="translating_to"
    >
      <Connect(TextInputElement)
        data-spec="translation-input"
        name="case-1"
        placeholder="English Translation"
        showLabel={false}
        value=""
      />
    </div>
  </div>
  <div
    className="translating_operator_container"
    data-spec="operator-field"
    key="translate_case-2"
  >
    <div
      className="translating_operator"
      data-spec="verbose-name"
    />
    <div
      className="translating_from"
      data-spec="argument-to-translate"
    >
      @run
    </div>
    <div
      className="translating_to"
    >
      <Connect(TextInputElement)
        data-spec="translation-input"
        name="case-2"
        placeholder="English Translation"
        showLabel={false}
        value=""
      />
    </div>
  </div>
</React.Fragment>
`;

exports[`SwitchRouterForm render should render exit translation form 1`] = `<div />`;
